File: articles/0004_std_lib_http_client/src/python/httppy/transport.py
from typing import Protocol

class Transport(Protocol):
    def connect(self, host: str, port: int) -> None:
        ...

    def write(self, data: bytes) -> int:
        ...

    def read_into(self, buffer: bytearray) -> int:
        ...

    def close(self) -> None:
        ...

File: articles/0004_std_lib_http_client/src/rust/src/transport.rs
use crate::error::Result;

pub trait Transport {
    fn connect(&mut self, host: &str, port: u16) -> Result<()>;

    fn write(&mut self, buf: &[u8]) -> Result<usize>;

    fn read(&mut self, buf: &mut [u8]) -> Result<usize>;

    fn close(&mut self) -> Result<()>;
}